### PR TITLE
Fix timepoint specific observables

### LINKED ISSUE #1214: Timepoint-specific noise parameters not detected
With @lcontento .

In a PEtab test model (attached), we use a noiseFormula `sqrt(alpha * noiseParameter1_obs1)` with measurement table

```tsv
observableId    simulationConditionId   time    measurement     noiseParameters
obs1    condition1      1.0     3       1.0
obs1    condition1      2.0     -5      0.25
```

Thus, we effectively have a time-dependent sigma. Had we used formula `sqrt(noiseParameter2_obs1 * noiseParameter1_obs1)` with replacement `1.0;alpha` and `0.25;alpha` in the measurement table, this would have been detected. Like this, however AMICI does not see any problem and simulates the model wrongly. Namely, 

```python
from pypesto.petab import PetabImporter
importer = PetabImporter.from_yaml('test.yaml')
objective = importer.create_problem().objective

print(objective(importer.petab_problem.x_nominal_free_scaled))
print(objective([4]))
print(objective([5]))

>>> 204.95158270528947
>>> 204.95158270528947
>>> 204.95158270528947
```

ignores any parameterization and implicitly uses another noise variable (which?).

**Suggestions**
1) Capture such cases and raise a warning (should be possible by evaluating the final noise formula after replacements)
2) Implement timepoint specific noise parameters :) .

### PR DESCRIPTION
**Pull request by FFroehlich**
fixes #1214

### CODE REVIEW

File: python/amici/petab_import.py [hunk 1]
```diff
@@ -574,35 +574,127 @@ def import_model_sbml(
 import_model = import_model_sbml
 
 
-def get_observation_model(observable_df: pd.DataFrame
-                          ) -> Tuple[Dict[str, Dict[str, str]],
-                                     Dict[str, str],
-                                     Dict[str, Union[str, float]]]:
+def has_timepoint_specific_measurement_overrides(measurement_df: pd.DataFrame):
```
Comment (dweindl) [inline python/amici/petab_import.py:577]: Could have used https://petab.readthedocs.io/en/stable/build/_autosummary/petab.lint.html#petab.lint.measurement_table_has_timepoint_specific_mappings

File: python/amici/petab_import.py [hunk 2]
```diff
@@ -574,35 +574,127 @@ def import_model_sbml(
...
+        updated observable + measurement table
     """
 
     if observable_df is None:
-        return {}, {}, {}
+        return dict(), dict(), dict()
+
+    if has_timepoint_specific_measurement_overrides(measurement_df):
+        raise ValueError(
+            'AMICI does not support importing models with  invidual '
```
Comment (dweindl) [inline python/amici/petab_import.py:683]:
```suggestion
            'AMICI does not support importing models with individual '
```

File: python/amici/petab_import.py [hunk 3]
```diff
@@ -574,35 +574,127 @@ def import_model_sbml(
...
+            'overrides for noise or observable  parameters. Please apply '
```
Comment (dweindl) [inline python/amici/petab_import.py:684]:
```suggestion
            'overrides for noise or observable parameters. Please apply '
```

File: python/amici/petab_import.py [hunk 4]
```diff
@@ -689,6 +780,10 @@ def parse_cli_args():
     parser.add_argument('--no-compile', action='store_false',
                         dest='compile',
                         help='Only generate model code, do not compile')
+    parser.add_argument('--flatten', dest='flatten', default=False,
+                        action='store_true',
+                        help='Flatten measurement specific overrides of '
+                             'observable and noise parameters')
```
Comment (dweindl) [inline python/amici/petab_import.py:786]: :+1:

File: python/amici/petab_import.py [hunk 5]
```diff
@@ -574,35 +574,127 @@ def import_model_sbml(
...
+        for field in [petab.NOISE_PARAMETERS, petab.OBSERVABLE_PARAMETERS]
+    ]
+    return has_timepoint_noise_overrides or has_timepoint_observable_overrides
+
+
+def flatten_timepoint_specific_overrides(problem: petab.Problem):
+    """
+    This function flattens out any measurement specific specifications of
+    noise or observable parameters by creating synthetic observables for
+    every unique occurence.
```
Comment (dweindl) [inline python/amici/petab_import.py:602]:
```suggestion
    every unique occurrence.
```

File: python/amici/petab_import.py [hunk 6]
```diff
@@ -574,35 +574,127 @@ def import_model_sbml(
...
+        len([x for x in measurement_df[field].unique()
+             if isinstance(x, str) and not re.match(nan_pat, x)
+             and not floatable(x)]) > 0
+        if field in measurement_df else False
+        for field in [petab.NOISE_PARAMETERS, petab.OBSERVABLE_PARAMETERS]
+    ]
+    return has_timepoint_noise_overrides or has_timepoint_observable_overrides
+
+
+def flatten_timepoint_specific_overrides(problem: petab.Problem):
```
Comment (dweindl) [inline python/amici/petab_import.py:598]: Looks like a better implementation than https://petab.readthedocs.io/en/stable/build/_autosummary/petab.core.html#petab.core.flatten_timepoint_specific_output_overrides. Maybe we can update the function there afterwards. There is already some test case.

**Review by dweindl — APPROVED** [review]

File: python/amici/petab_import.py [hunk 7]
```diff
@@ -574,10 +575,45 @@ def import_model_sbml(
 import_model = import_model_sbml
 
 
-def get_observation_model(observable_df: pd.DataFrame
-                          ) -> Tuple[Dict[str, Dict[str, str]],
-                                     Dict[str, str],
-                                     Dict[str, Union[str, float]]]:
+def has_timepoint_specific_measurement_overrides(measurement_df: pd.DataFrame):
+
+    def unfloatable(x: str) -> bool:
```
Comment (yannikschaelte) [inline python/amici/petab_import.py:580]:
```suggestion
    def unfloatable(x: str) -> bool:
    """Determine whether `x` can be converted to float(s).
    
    `x` is a ";" separated list of numerics and parameter identifiers.
    """
    or sth like that
```

File: python/amici/petab_import.py [hunk 8]
```diff
@@ -574,10 +575,45 @@ def import_model_sbml(
...
+            df = df[df[field].apply(unfloatable)]
+
+    if df.empty:
+        return False
+
+    groupvars = [petab.OBSERVABLE_ID] + \
+                [x for x in [petab.SIMULATION_CONDITION_ID,
+                             petab.PREEQUILIBRATION_CONDITION_ID]
+                 if x in df]
+
```
Comment (yannikschaelte) [inline python/amici/petab_import.py:605]:
```suggestion

# data frame has timepoint specific overrides if grouping by noise parameters
#  and observable parameters in addition to observable, condition and preeq id
#  yields more groups
```

File: .github/workflows/test_petab_test_suite.yml [hunk 9]
```diff
@@ -50,6 +50,9 @@ jobs:
       - name: Install python package
         run: |
           scripts/installAmiciSource.sh
+      - name: Temp fix petab
+          run: |
```
Comment (dweindl) [inline .github/workflows/test_petab_test_suite.yml:54]:
```suggestion
        run: |
```

File: python/sdist/setup.cfg [hunk 10]
```diff
@@ -41,7 +41,7 @@ include_package_data = True
 zip_safe = False
 
 [options.extras_require]
-petab = petab>=0.1.11
+petab = petab>=0.1.16
```
Comment (dweindl) [inline python/sdist/setup.cfg:44]:
```suggestion
petab = petab>=0.1.17
```

**Pull request merged**